tianshou/policy/modelfree/qrdqn.py

```diff
@@ -0,0 +1,92 @@
+
+
+class QRDQNPolicy(DQNPolicy):
+    """Implementation of Quantile Regression Deep Q-Network. arXiv:1710.10044.
+
+    :param torch.nn.Module model: a model following the rules in
+        :class:`~tianshou.policy.BasePolicy`. (s -> logits)
+    :param torch.optim.Optimizer optim: a torch.optim for optimizing the model.
+    :param float discount_factor: in [0, 1].
+    :param int num_quantiles: the number of quantile midpoints in the inverse
+        cumulative distribution function of the value, defaults to 200.
+    :param int estimation_step: greater than 1, the number of steps to look
+        ahead.
+    :param int target_update_freq: the target network update frequency (0 if
+        you do not use the target network).
+    :param bool reward_normalization: normalize the reward to Normal(0, 1),
+        defaults to False.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.DQNPolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        discount_factor: float = 0.99,
+        num_quantiles: int = 200,
+        estimation_step: int = 1,
+        target_update_freq: int = 0,
+        reward_normalization: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(model, optim, discount_factor, estimation_step,
+                         target_update_freq, reward_normalization, **kwargs)
+        assert num_quantiles > 1, "num_quantiles should be greater than 1"
+        self._num_quantiles = num_quantiles
+        tau = torch.linspace(0, 1, self._num_quantiles + 1)
+        self.tau_hat = torch.nn.Parameter(
+            (tau[:-1] + tau[1:]) / 2, requires_grad=False)
+
+    def _target_q(
+        self, buffer: ReplayBuffer, indice: np.ndarray
+    ) -> torch.Tensor:
+        batch = buffer[indice]  # batch.obs_next: s_{t+n}
+        if self._target:
+            a = self(batch, input="obs_next").act
+            next_dist = self(
+                batch, model="model_old", input="obs_next"
+            ).logits
+        else:
+            next_b = self(batch, input="obs_next")
+            a = next_b.act
+            next_dist = next_b.logits
+        next_dist = next_dist[np.arange(len(a)), a, :]
+        return next_dist  # shape: [bsz, num_quantiles]
+
+    def compute_q_value(self, logits: torch.Tensor) -> torch.Tensor:
+        """Compute the q value based on the network's raw output logits."""
+        return logits.mean(2)
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        if self._target and self._iter % self._freq == 0:
+            self.sync_weight()
+        self.optim.zero_grad()
+        weight = batch.pop("weight", 1.0)
+        curr_dist = self(batch).logits
+        act = batch.act
+        curr_dist = curr_dist[np.arange(len(act)), act, :]
+        target_dist = batch.returns
+        u = target_dist.unsqueeze(1) - curr_dist.unsqueeze(2)
```

Comment:
Why did you broadcast here @shengxiang19? I want to use `F.smooth_l1_loss` but get the following message here:

```
/home/trinkle/github/tianshou-new/tianshou/policy/modelfree/qrdqn.py:81: UserWarning: Using a target size (torch.Size([64, 200, 1])) that is different to the input size (torch.Size([64, 1, 200])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
```

Reply:
It's hard to explain. You can see (8) in QR-DQN paper.  I think F.smooth_l1_loss can't meet our demand.

Reply:
For each element in curr_dist, I want to calculate the difference between it and each element in target_dist.

Reply:
How about this https://github.com/thu-ml/tianshou/pull/276/commits/7ab816034c74d4bfa672a495074956265dee81f9
Actually `torch.where` is much slower than `F.smooth_l1_loss`, after this modification, the training speed increased from 988/s to 1280/s in my own computer. BTW please check if my modification is equivalent.

Reply:
I think it is ok.